### PR TITLE
Cluster timeout

### PR DESCRIPTION
**Pull request by BLasan**
## Purpose
Configure separate timeout values for backends instead of configuring a default timeout values for routes.

## Approach
- Introduce a new property `timeout` in Backend CR
- Override the `Timeout` & `IdleTimeout` values in Endpoint Config with the defined value. Otherwise, default values are used.
- Eliminate Cluster Timeout getting overridden as it represents a timeout for a new connection to the cluster.
- New Backend CR looks as follows.
```
apiVersion: dp.wso2.com/v1alpha1
kind: Backend
metadata:
  name: infra-backend-v1
  namespace: gateway-integration-test-infra
spec:
  timeout:
    maxRouteTimeoutSeconds: 60
    routeIdleTimeoutSeconds: 400
    routeTimeoutSeconds: 40
  services:
  - host: infra-backend-v1.gateway-integration-test-infra
    port: 8080
---
apiVersion: dp.wso2.com/v1alpha1
kind: Backend
metadata:
  name: infra-backend-v2
  namespace: gateway-integration-test-infra
spec:
  services:
  - host: infra-backend-v2.gateway-integration-test-infra
    port: 8080
```

Enforcer Response Headers
```
ok_response {
  headers {
    header {
      key: "x-envoy-upstream-rq-timeout-ms"
      value: "40000"
    }
  }
  headers {
    header {
      key: "Authorization"
      value: "Basic YWRtaW4KOmFkbWluCg=="
    }
  }
  headers {
    header {
      key: "x-wso2-cluster-header"
      value: "a3b58ccf-6ecc-4557-b5bb-0a35cce38256__prod.gw.wso2.com_http-bin-api1.0.8_5c1a1d01-be14-4d77-803c-33ae79eeb4810"
    }
  }
  headers_to_remove: "internal-key"
  headers_to_remove: "x-wso2-client-certificate"
}
```
Route Config
```
                     "cluster_header": "x-wso2-cluster-header",
                      "auto_host_rewrite": true,
                      "idle_timeout": "400s",
                      "upgrade_configs": [
                        {
                          "upgrade_type": "websocket",
                          "enabled": false
                        }
                      ],
                      "regex_rewrite": {
                        "pattern": {
                          "regex": "^/http-bin-api/1\\.0\\.8/hello((?:/.*)*)"
                        },
                        "substitution": "/hello\\1"
                      }
                    },
```
### Related Issues
- https://github.com/wso2/apk/issues/1279

### CODE REVIEW

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 1]
```diff
@@ -105,6 +105,8 @@ func CreateRoutesWithClusters(adapterInternalAPI model.AdapterInternalAPI, inter
 	apiVersion := adapterInternalAPI.GetVersion()
 
 	conf := config.ReadConfigs()
+
+	// Get the timeout from the default config as this is used for the enforcer timeout
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:109]: enforcer timeout? what does it mean? there's no usage for enforcer of a cluster timeout

File: adapter/internal/oasparser/model/adapter_internal_api.go [hunk 2]
```diff
@@ -439,6 +440,7 @@ func (endpointCluster *EndpointCluster) validateEndpointCluster() error {
 			}
 			// Validate timeout
 			conf := config.ReadConfigs()
+			// Set timeout to default if not provided
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/model/adapter_internal_api.go:443]: is this a TODO?

File: adapter/internal/oasparser/model/http_route.go [hunk 3]
```diff
@@ -80,6 +83,9 @@ func (swagger *AdapterInternalAPI) SetInfoHTTPRouteCR(httpRoute *gwapiv1b1.HTTPR
 		var resourceRatelimitPolicy *dpv1alpha1.RateLimitPolicy
 		hasPolicies := false
 		var scopes []string
+		var timeoutInMillis uint32 = config.Envoy.Upstream.Timeouts.RouteTimeoutInSeconds * 1000
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/model/http_route.go:86]: shall we remove this from the conf as route timeouts are coming from backend CR

Comment (BLasan) [reply adapter/internal/oasparser/model/http_route.go:86]: @AmaliMatharaarachchi What if the `Backend` doesn't have the `timeout` property defined? If it's nil, then we need to have some default values to be set. Or else we have to omit configuring the `EndpointCluster` with the idle timeout and timeout properties which won't add the envoy headers to the enforcer response. https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-retry-on

Comment (AmaliMatharaarachchi) [reply adapter/internal/oasparser/model/http_route.go:86]: they still would have the default values set by envoy internally, right?

Comment (BLasan) [reply adapter/internal/oasparser/model/http_route.go:86]: Yes, they would consider their default values. I will then conditionally configure the endpoint cluster config for the timeouts

File: adapter/internal/operator/apis/dp/v1alpha1/backend_webhook.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//+kubebuilder:webhook:path=/validate-dp-wso2-com-v1alpha1-backend,mutating=false,failurePolicy=fail,sideEffects=None,groups=dp.wso2.com,resources=backends,verbs=create;update,versions=v1alpha1,name=vbackend.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &Backend{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *Backend) ValidateCreate() error {
+	return r.validateBackendSpec()
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/operator/apis/dp/v1alpha1/backend_webhook.go:60]: need to add this to the `ValidateUpdate` too

File: adapter/internal/operator/apis/dp/v1alpha1/backend_webhook.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil
+}
+
+func (r *Backend) validateBackendSpec() error {
+	var allErrs field.ErrorList
+	timeout := r.Spec.Timeout
+	if timeout != nil {
+		if timeout.MaxRouteTimeoutSeconds < timeout.RouteTimeoutSeconds {
```
[the remark this answers is not in the document]
Comment (BLasan) [reply adapter/internal/operator/apis/dp/v1alpha1/backend_webhook.go:83]: `IdleTimeout` and `RouteTimeout` are focusing on two different aspects and we don't have to validate against each other

File: adapter/internal/oasparser/model/adapter_internal_api.go [hunk 6]
```diff
@@ -438,13 +438,6 @@ func (endpointCluster *EndpointCluster) validateEndpointCluster() error {
 			if endpointCluster.Config.RetryConfig != nil {
 				endpointCluster.Config.RetryConfig.validateRetryConfig()
 			}
-			// Validate timeout
-			conf := config.ReadConfigs()
```
[the remark this answers is not in the document]
Comment (BLasan) [reply adapter/internal/oasparser/model/adapter_internal_api.go:441]: Removed from the default config. https://github.com/wso2/apk/pull/1313/files#diff-0b437337d1166a184d4184f4f3763fa5eadb3190e3a9de46f2406fffa3d0c4ceL67

File: adapter/internal/operator/apis/dp/v1alpha1/backend_webhook.go (hunk 4 above)
Comment (BLasan) [reply adapter/internal/operator/apis/dp/v1alpha1/backend_webhook.go:60]: https://github.com/wso2/apk/pull/1313/files#diff-d2df7698f14faaf6890fb6beffbd51011341c1b2a51ffb35962fe9848180b944R64

File: adapter/internal/oasparser/model/adapter_internal_api.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (AmaliMatharaarachchi) [reply adapter/internal/oasparser/model/adapter_internal_api.go:441]: you haven't removed unused variables. check in types.go

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Pull request merged**